From b0497ca4b508d6f894d78b16e0e06616a2a36c16 Mon Sep 17 00:00:00 2001
From: Chris Morgan <macromorgan@hotmail.com>
Date: Fri, 24 Feb 2023 13:33:20 -0600
Subject: [PATCH] ucm2: Rockchip: rk817: Add ALSA UCM support

Add ALSA-UCM support for the Rockchip RK817 audio codec. This codec
is typically configured either with an internal or external amplifier
as reflected by the longname. This configuration has been tested on
the Anbernic RG353P (rk817_ext) and the Odroid Go Advance (rk817_int).

Changes from V1:
 - Use a constant to control if the output mux is set to HP or SPK.

Link: https://lore.kernel.org/alsa-devel/20230224193320.1503-1-macroalpha82@gmail.com/
Signed-off-by: Chris Morgan <macromorgan@hotmail.com>
Signed-off-by: Jaroslav Kysela <perex@perex.cz>
---
Upstream-Status: Backport from 1.2.9

 ucm2/Rockchip/rk817-sound/HiFi.conf        | 68 ++++++++++++++++++++++
 ucm2/Rockchip/rk817-sound/rk817-sound.conf |  6 ++
 ucm2/conf.d/simple-card/rk817_ext.conf     |  1 +
 ucm2/conf.d/simple-card/rk817_int.conf     |  1 +
 4 files changed, 76 insertions(+)
 create mode 100644 ucm2/Rockchip/rk817-sound/HiFi.conf
 create mode 100644 ucm2/Rockchip/rk817-sound/rk817-sound.conf
 create mode 120000 ucm2/conf.d/simple-card/rk817_ext.conf
 create mode 120000 ucm2/conf.d/simple-card/rk817_int.conf

diff --git a/ucm2/Rockchip/rk817-sound/HiFi.conf b/ucm2/Rockchip/rk817-sound/HiFi.conf
new file mode 100644
index 00000000..8d905678
--- /dev/null
+++ b/ucm2/Rockchip/rk817-sound/HiFi.conf
@@ -0,0 +1,68 @@
+Define.pbk_mux "SPK"
+
+If.1 {
+	Condition {
+		Type ControlExists
+		Control "name='Internal Speakers Switch'"
+	}
+
+	True {
+		Define.pbk_mux "HP"
+		SectionDevice."Speaker".EnableSequence [
+			cset "name='Internal Speakers Switch' on"
+		]
+
+		SectionDevice."Speaker".DisableSequence [
+			cset "name='Internal Speakers Switch' off"
+		]
+	}
+}
+
+SectionDevice."Speaker" {
+	Comment "Internal Speaker"
+
+	EnableSequence [
+		cset "name='Playback Mux' ${var:pbk_mux}"
+	]
+
+	Value {
+		PlaybackMixerElem "Master Playback Volume"
+		PlaybackPriority 100
+		PlaybackPCM "hw:${CardId}"
+	}
+
+	ConflictingDevice [
+		"Headphones"
+	]
+}
+
+SectionDevice."Mic" {
+	Comment "Microphone"
+
+	Value {
+		CapturePriority 100
+		CapturePCM "hw:${CardId}"
+		CaptureMixerElem "Mic Capture Gain"
+		CaptureMasterElem "Master Capture Volume"
+	}
+}
+
+SectionDevice."Headphones" {
+	Comment "Headphones"
+
+	EnableSequence [
+		cset "name='Playback Mux' HP"
+	]
+
+	Value {
+		PlaybackMixerElem "Master Playback Volume"
+		PlaybackPriority 200
+		PlaybackPCM "hw:${CardId}"
+		JackControl "Headphones Jack"
+		JackHWMute "Speaker"
+	}
+
+	ConflictingDevice [
+		"Speaker"
+	]
+}
diff --git a/ucm2/Rockchip/rk817-sound/rk817-sound.conf b/ucm2/Rockchip/rk817-sound/rk817-sound.conf
new file mode 100644
index 00000000..90ec7946
--- /dev/null
+++ b/ucm2/Rockchip/rk817-sound/rk817-sound.conf
@@ -0,0 +1,6 @@
+Syntax 4
+
+SectionUseCase."HiFi" {
+	File "/Rockchip/rk817-sound/HiFi.conf"
+	Comment "Play HiFi quality music"
+}
diff --git a/ucm2/conf.d/simple-card/rk817_ext.conf b/ucm2/conf.d/simple-card/rk817_ext.conf
new file mode 120000
index 00000000..6fa8db7d
--- /dev/null
+++ b/ucm2/conf.d/simple-card/rk817_ext.conf
@@ -0,0 +1 @@
+../../Rockchip/rk817-sound/rk817-sound.conf
\ No newline at end of file
diff --git a/ucm2/conf.d/simple-card/rk817_int.conf b/ucm2/conf.d/simple-card/rk817_int.conf
new file mode 120000
index 00000000..6fa8db7d
--- /dev/null
+++ b/ucm2/conf.d/simple-card/rk817_int.conf
@@ -0,0 +1 @@
+../../Rockchip/rk817-sound/rk817-sound.conf
\ No newline at end of file